arch-split: InvariantUpdates_H and StateRelation

Xaphiosis opened the pull request:
Please review on a commit-by-commit basis. The general scheme is I move the arch specific files in one commit, then do the arch split in the next commit, and if updates are more complex/interesting, they end up with more commits.

Also note: some of the `FIXME arch-split` comments are interesting for this review, but should probably be chucked out or converted to normal comments when we actually want to merge this.

Interesting things, mostly for StateRelation:

Similarly to invariants, the state relation contains a large number of
definitions. To reduce load on locales and number of theory files, the
architecture-specific definitions go first (ArchStateRelation), then the
generic definitions and lemmas (StateRelation), and finally the
arch-specific lemmas (ArchStateRelationLemmas).

- due to ghost_relation being arch-specific on AARCH64, ghost_relation
  becomes arch-specific, with a generic `ghost_relation_wrapper`
  exported into generic context
- ghost_relation is required to not depend on machine state
- the [a]obj_relation_cuts infrastructure conflated arch and generic
  object cases in other_obj_relation and those had to be split up
- obj_relation_cutsE unfortunately still spills all the cases, but
  defining that properly can only happen when dealing with KHeap_R

Xaphiosis commented:
Currently broken due to conflicts with master, need to rebase and fix.

lsf37 commented on proof/refine/InvariantUpdates_H.thy:
```diff
@@ -0,0 +1,457 @@
+  valid_arch_tcb_lift'
+  tcb_at_cte_at'
+  gen_objBitsT_simps
+  objBitsT_koTypeOf (* FIXME arch-split: consider declaring [simp] here rather than Schedule_R *)
```
That might be nice to do, but I'd leave it for later, so that the fallout from `[simp]` isn't mixed in with the rest. I.e. leave in the FIXME for now.

Xaphiosis replied:
Agreed, I think it'll also be easier to fix the fallout when things are more arch-split. Currently having to go do a fix 5 times is annoying.

lsf37 commented on proof/refine/InvariantUpdates_H.thy:
```diff
@@ -0,0 +1,457 @@
+  by (simp add: ct_idle_or_in_cur_domain'_def tcb_in_cur_domain'_def)
+
+(* FIXME arch-split: export? *)
+lemma invs'_machine:
```
I don't think I understand the comment. We're in a generic file at the top level, so where would we export to?

Xaphiosis replied:
I think it's a leftover from when it was in the Arch locale, forgot to remove comment when I moved it.

- [ ] remove `FIXME arch-split: export?` comments here and for `invs_no_cicd'_machine`

lsf37 commented on proof/refine/AARCH64/ArchStateRelation.thy:
```diff
@@ -6,53 +6,30 @@
-  "cte_map \<equiv> \<lambda>(oref, cref). oref + (of_bl cref << cte_level_bits)"
-
-lemmas cte_map_def' = cte_map_def[simplified cte_level_bits_def shiftl_t2n mult_ac, simplified]
+(* FIXME arch-split: these are generic, but due to many definitions, the Arch theory goes before
+   the generic one, so there's nowhere sensible to move to *)
+type_synonym obj_relation_cut = "Structures_A.kernel_object \<Rightarrow> Structures_H.kernel_object \<Rightarrow> bool"
+type_synonym obj_relation_cuts = "(machine_word \<times> obj_relation_cut) set"
```
A small `_pre` theory?

Xaphiosis replied:
I tried to avoid creating yet another pre theory given it only needed a few lines. I'm mindful that pretty much every theory is going to become two, and Invariants_H already became 4 theories, so to have this one be 4 as well... if you think that's an improvement, I'll do it.

Xaphiosis replied:
this is done now

lsf37 commented on proof/refine/AARCH64/ArchStateRelation.thy:
```diff
@@ -289,133 +122,22 @@ definition is_other_obj_relation_type :: "a_type \<Rightarrow> bool" where
+(* FIXME arch-split: provided only so that ghost_relation can be used within generic definition
+   of state_relation (since arch_state_relation doesn't have access to kheap, and
+   gsPTTypes on AARCH64 isn't generic) *)
+definition ghost_relation_wrapper :: "det_state \<Rightarrow> kernel_state \<Rightarrow> bool" where
```
do we want it to be `det_state`? it looks like it would be fine the generic state

Xaphiosis replied:
Turns out on master all `state_relation` takes `det_state` already, and that's where I got it from. After discussion, we both decided that indicates no need for generalisation here.

lsf37 commented on proof/refine/AARCH64/ArchStateRelation.thy:
```diff
@@ -427,78 +149,9 @@ definition arch_state_relation :: "(arch_state \<times> AARCH64_H.kernel_state)
+(* FIXME arch-split: object_type is arch-specific with one generic field, inverse of usual setup,
+   but if we split off the generic part, it would have to go before arch definitions, which would
+   require more locales or an extra generic theory file *)
 definition APIType_map :: "Structures_A.apiobject_type \<Rightarrow> AARCH64_H.object_type" where
```
I wouldn't be opposed to a small extra theory if it makes the definitions nicer. We don't want that everywhere in the proofs, but for invariants and state relation I'd say it's worth it.

corlewis replied:
Do we want to rename `APIType_map2` and move it here (or to a generic file if it can be split)? That could probably be done in a separate PR, except that we would lose track of this location and the connection to the definition that you're removing.

lsf37 replied:
Agree, I would do that when we hit Retype_R.

Xaphiosis replied:
When we get to Retype_R, we'll be forced to arch-split APIType_map2, which would be a good point to rename it. Given the grief it gave me today, I'm not likely to forget it.

lsf37 commented on proof/refine/AARCH64/ArchStateRelationLemmas.thy:
```diff
@@ -0,0 +1,176 @@
+  by (simp add: obj_relation_cuts_def2 a_type_def well_formed_cnode_n_def length_set_helper
+           split: Structures_A.kernel_object.split AARCH64_A.arch_kernel_obj.split)
+
+(* FIXME arch-split: split off generic? *)
+lemma obj_relation_cutsE:
```
This is one of the more fundamental questions. If we have case distinction lemmas like this in here, that means that all lemmas that use them must be arch specific. That's not necessarily a problem, but it is a potential point for optimisation later.

I'd vote for leaving the FIXME in for now and see how it goes with just the arch-specific version of it. If it draws in a lot of later lemmas that we really would like to be generic, then we should come back here and split it up into generic cases and arch cases. Doing it immediately feels like premature optimisation.

lsf37 commented on proof/refine/AARCH64/ArchStateRelationLemmas.thy:
```diff
@@ -0,0 +1,176 @@
+                     AARCH64_A.arch_kernel_obj.splits)
+  done
+
+(* FIXME arch-split: export? *)
+lemma is_other_obj_relation_type_CapTable:
+  "\<not> is_other_obj_relation_type (ACapTable n)"
+  by (simp add: is_other_obj_relation_type_def)
+
+(* FIXME arch-split: export? *)
+lemma is_other_obj_relation_type_TCB:
```
Yes, those look both worthy of export

Xaphiosis replied:
- [ ] export these two lemmas

lsf37 commented on proof/refine/StateRelation.thy:
```diff
@@ -0,0 +1,444 @@
+   (case (obj, obj') of
+      (Endpoint ep, KOEndpoint ep') \<Rightarrow> ep_relation ep ep'
+    | (Notification ntfn, KONotification ntfn') \<Rightarrow> ntfn_relation ntfn ntfn'
+    | _ \<Rightarrow> False)"
```
Was probably there before, but the outer parentheses around `case` are not needed (only for `=`, not for `==`)

Xaphiosis replied:
done

Xaphiosis commented:
Hopefully addressed all comments as separate commits. Please take another look while the regression tests run and make sure I didn't accidentally mess something up.

corlewis commented on proof/refine/ARM/ArchInvariantUpdates_H.thy:
```diff
@@ -5,482 +5,23 @@
+lemma valid_arch_state'_interrupt[simp, InvariantUpdates_H_assms]:
   "valid_arch_state' (ksInterruptState_update f s) = valid_arch_state' s"
-  by (simp add: valid_arch_state'_def)
+  by (simp add: valid_arch_state'_def cong: option.case_cong)
```
Oh, is it to make it consistent with the other architectures?

corlewis commented on proof/refine/InvariantUpdates_H.thy:
```diff
@@ -0,0 +1,457 @@
+
+lemma valid_bitmaps_ksSchedulerAction_update[simp]:
+   "valid_bitmaps (ksSchedulerAction_update f s) = valid_bitmaps s"
+   by (simp add: valid_bitmaps_def bitmapQ_defs)
```
Do we care about fixing style while moving these lemmas? If we do, this is overindented.

Xaphiosis replied:
Fixed, thanks.

lsf37 commented on proof/refine/X64/ArchStateRelation.thy:
```diff
@@ -0,0 +1,202 @@
+imports StateRelationPre
+begin
+
+(* FIXME arch-split: these are generic, but due to many definitions, the Arch theory goes before
+   the generic one, so there's nowhere sensible to move to *)
+text \<open>
+  A pspace and a tree are related if every object in the pspace
+  corresponds to an object in the tree. Some abstract objects
+  like CapTables correspond to multiple concrete ones, thus we
+  have to make cuts.\<close>
+type_synonym obj_relation_cut = "Structures_A.kernel_object \<Rightarrow> Structures_H.kernel_object \<Rightarrow> bool"
+type_synonym obj_relation_cuts = "(machine_word \<times> obj_relation_cut) set"
```
also here

lsf37 commented on proof/refine/ARM/ArchStateRelation.thy:
```diff
@@ -0,0 +1,174 @@
+imports StateRelationPre
+begin
+
+(* FIXME arch-split: these are generic, but due to many definitions, the Arch theory goes before
+   the generic one, so there's nowhere sensible to move to *)
+text \<open>
+  A pspace and a tree are related if every object in the pspace
+  corresponds to an object in the tree. Some abstract objects
+  like CapTables correspond to multiple concrete ones, thus we
+  have to make cuts.\<close>
+type_synonym obj_relation_cut = "Structures_A.kernel_object \<Rightarrow> Structures_H.kernel_object \<Rightarrow> bool"
+type_synonym obj_relation_cuts = "(machine_word \<times> obj_relation_cut) set"
```
one more

lsf37 reviewed (approved):
Looks good, but you still need to remove the StateRelationPre definitions from the other architectures. StateRelationPre is now smaller than we had planned, but I think it's still Ok to have it. We might find more things in the future that we want to put there.

Xaphiosis commented on proof/refine/AARCH64/CSpace1_R.thy:
```diff
@@ -6659,6 +6659,7 @@ lemma cteSwap_corres:
          (cap_swap scap src dcap dest) (cteSwap scap' src' dcap' dest')"
   (is "corres _ ?P ?P' _ _") using assms including no_pre
   supply None_upd_eq[simp del]
+  supply AARCH64.ghost_relation_wrapper_def[simp] (* FIXME arch-split *)
```
Reviewed. There was a spurious one in ArchAcc on AARCH64, but otherwise:
* CSpace1_R.thy - needed in cteSwap_corres (not in Arch)
* Retype_R.thy - currently needed for state_relationE expansion
* Detype_R.thy - currently needed for state_relationE expansion
* Untyped_R.thy - needed for updateFreeIndex_corres (not in Arch)

Xaphiosis commented on proof/refine/StateRelation.thy:
```diff
@@ -0,0 +1,444 @@
+  apply (fastforce dest: pspace_relation_absD)
+  done
+
+lemma eq_trans_helper:
+  "\<lbrakk> x = y; P y = Q \<rbrakk> \<Longrightarrow> P x = Q"
+  by simp
```
added comment: `(* used for generating architecture-specific cap_relation split lemmas *)` for now, let me know what you think

Xaphiosis commented:
Addressed @corlewis's review comments and the derp @lsf37 pointed out. Hopefully that's all of them, let me know if there's anything else.

lsf37 commented:
Good to go from my side!
